### PR TITLE
New path tree

### PR DESCRIPTION
**Pull request by dwhswenson**
This PR starts on a new path tree code. The main reason for this is that the old path trees do not fully work with SimStore because they rely on a bad design decision that SimStore is removing. Additionally, the old path trees don't work with two-way shooting in either SimStore or netcdfplus.

Furthermore, there is no unit testing for the old path trees, and they're written in high complexity code that would be difficult to test. There's also very little documentation to help anyone else maintain that code.

Plus, I think the new version has a much nicer API, especially for the most common use case (selecting replica 0):

```python
tree = PathTree(steps)
tree.matplotlib()
```

Also, the API for added new plotting utilities is very straightforward: you create a class that implements two functions:

* `draw_trajectory(row, step, options)`
* `draw_connector(x, bottom, top, step, options)`

In these `step` is a namedtuple that carries the `mover`, the `trajectory`, the bool `accepted`, and `segments`, which provides information about which parts of the trajectory were new and which existed in the prior trajectory. `options` us the object which carries all the user-customizable plotting options, also contains a function that will get the left and right endpoints of the trajectory. `row` is the row number of the trajectory. `x` is the location of the shooting point, and bottom and top are row numbers to connect the trajectory to the shooting point origin. Examples of this are in the `mpl.py` file.

Unfortunately, there's a lot of really good functionality in the old path trees that will take time to replicate. I'm curious if anyone has ever actually used some of those (like changing the CSS to use different colors, especially colors that vary across a trajectory, representing the values of the snapshot, or adding text labels per snapshot).

---

2024 UPDATE: Overview of the contents here:

* `path_tree.py`: Generate the objects that describe the path tree. This is content, as separated from presentation, but has conveniences for a few presentation modes.
* `options.py`: Attempt to generalize presentation options in a way that might be valid across multiple presentation tools. This is the part of the API that is most likely to undergo massive changes.
* `plotter.py`: ABC for presentation-specific plotting classes. This may see some future refactoring.
* `mpl.py`: Tools for using matplotlib for presentation.
* `svg.py`: Tools for using SVG for presentation.

The use modes that are expected to be supported are:

1. Use conveniences in the `PathTree` object to plot:

    ```python
    tree = PathTree(steps)
    tree.matplotlib()
    # or
    tree.svg()
    ```

2. Explicitly do content and presentation in separate steps:

    ```python
    # content for any presentation mode
    tree = PathTree(steps)
    # matplotlib presentation
    mpl_plotter = mpl.ThinLines()
    mpl_plotter.draw(tree)
    # svg presentation
    svg_plotter = SVGLines()
    svg_plotter.draw(tree)
    ```

### CODE REVIEW

**Comment by sroet** [comment]
> Unfortunately, there's a lot of really good functionality in the old path trees that will take time to replicate. I'm curious if anyone has ever actually used some of those (like changing the CSS to use different colors, especially colors that vary across a trajectory, representing the values of the snapshot, or adding text labels per snapshot).

I believe the most I have ever done with the tree representations was [this least changed path notebook](https://gitlab.e-cam2020.eu/sroet/least_changed_path/-/blob/master/examples/Least_changed_path_example.ipynb) (It is old enough to have the  wrong colors xD) (for context it was mainly used to check that all the right frames were selected)

**Comment by dwhswenson** [comment]
I'm going to mark this as ready for review, because this needs to be in the 1.7 release (due out this weekend). The policy on "experimental" modules is that we don't require unit tests in place (even though it pains me not to have them). I've given this a decent bit of testing via a few Jupyter notebooks, which I tossed into a gist:

The API is not completely solidified yet, especially around how to configure more custom coloring options. However, I'm reasonably confident in the parts of the API the separate the content of the path tree from its presentation as SVG, Matplotlib, or other front-end.

Things I might try to add before merge:

* initial unit tests
* modifications to give defaults that distinguish rejected paths from accepted paths

Things we'd like to add in the future:

* side bar showing MC step number and decorrelated trajectories
* hover info in SVG

Given the lack of tests and reasonably large code base, I'm going to leave this open for review as long as I can, while still getting 1.7 released. This is ready for review and comment. Without review, I will merge this after at least 80 hours, no sooner than Fri 30 Aug 17:00 GMT (19:00 my local).

File: openpathsampling/experimental/path_tree/path_tree.py
Comment (sroet) [inline openpathsampling/experimental/path_tree/path_tree.py:249]: shouldn't this be brought inline with line `234`?
```suggestion
        plotter.draw(self)
```

File: openpathsampling/experimental/path_tree/svg.py
Comment (sroet) [inline openpathsampling/experimental/path_tree/svg.py:28]: leftover debug code?

**Review by sroet — CHANGES_REQUESTED** [review]
One spotted bug, one consideration, one simplification and some leftover commented out functionality

File: openpathsampling/experimental/path_tree/svg.py
Comment (dwhswenson) [reply openpathsampling/experimental/path_tree/svg.py]: refactored into a superclass, actually (same difference!) Removed in [16ede3b](https://github.com/openpathsampling/openpathsampling/pull/1048/commits/16ede3b5a486af0fff482dbb7f92382a90f23f57)

File: openpathsampling/experimental/path_tree/options.py
Comment (dwhswenson) [reply openpathsampling/experimental/path_tree/options.py:75]: yeah, I think this is a case where `allowed_movers` may default to `None` instead of `[]`, so this suggested is what was intended.

File: openpathsampling/experimental/path_tree/options.py
Comment (dwhswenson) [reply openpathsampling/experimental/path_tree/options.py:30]: TBH, I don't now remember what my thoughts on doing this as a defaultdict were. At this point, I have a strong suspicion the `options` will be heavily rewritten (now that I have 3 implementations to work with, I think it'll be easier to find commonalities).

To a large extent, the problem with my first pass of options is that I was trying to unify the efficiency of simple, single lines for each trajectory vs. boxes (that could, e.g., be difference colors for every frame). My view now is that it trying to be too clever, and we should just see those as two different families of path tree.

In any case, I'll leave this as is until the rewrite. There's a reason this part of the API is experimental!

File: openpathsampling/experimental/path_tree/path_tree.py
Comment (dwhswenson) [reply openpathsampling/experimental/path_tree/path_tree.py:57]: looking into this further; current code is working, but this makes me confused as to why (maybe this code isn't hit?) Is looks to me like that should be `blocks.append(...)`, which would also make resetting the `block` (local variable) make sense.

Comment (dwhswenson) [reply openpathsampling/experimental/path_tree/path_tree.py]: cleaned it up a bit, but this code isn't used yet (which explains why it wasn't breaking anything)

**Comment by dwhswenson** [comment]
@sroet : This can take another pass, if you have time. I think I addressed all the changes you requested.

Without review, I will merge this after at least 8 hours, no sooner than Sat 31 Aug 21:00 GMT (23:00 my local).

**Review by sroet — CHANGES_REQUESTED** [review]
One leftover doc that points to a function that doesn't seem to exist, LGTM otherwise!

File: openpathsampling/experimental/path_tree/mpl.py
Comment (dwhswenson) [reply openpathsampling/experimental/path_tree/mpl.py]: Good catch: that whole paragraph is now outdated. Fixed in 
f4b714f; if you have a chance, please mark an "approve" on this so I don't need to merge with override of requested changes!

**Review by sroet — APPROVED** [review]
LGTM

**Pull request merged**
